Cmake new multiconfig 2

mcharleb opened the pull request.

mcharleb committed:
Added support for multiple config files per OS

New build configs are in:

cmake/${OS}/${OS}-${BOARD}-${TARGET_LABEL}

i.e. posix-sitl-simple

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
Adding more CMakeLists.txt for posix build

src/systemcmds/topic_listener/CMakeLists.txt is
broken as it needs to generate a file

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb commented:
Don't merge this yet, just giving you visibility into what I have changed.

mcharleb committed:
Added more cmake support for posix build

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb commented on src/lib/mathlib/CMakeLists.txt:
```diff
@@ -0,0 +1,58 @@
+set_target_properties(${module} PROPERTIES
+        COMPILE_FLAGS "-Os")
+
+# FIXME - Not cmake compliant
```
@jgoppert I'm not sure what the rule below (from module.mk) is trying to do.

jgoppert replied:
I'm not sure either. @LorenzMeier, any insights?

mcharleb commented on src/systemcmds/topic_listener/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+set(stack 1800)
+set(main listener_main)
+
+# FIXME
```
@modevico How can the topic_listener.cpp be generated in the build tree in the appropriate location?

jgoppert replied:
If you generate a file in a subdirectory, it will be generated in the matching directory in the build directory, then if you need to include the source you can add ${CMAKE_CURRENT_BINARY_DIR}/file.cpp to the sources.

mcharleb replied:
@jgoppert Thanks! Its fixed now.

mcharleb committed:
More cmake support for posix build

topic_listener is disabled for now as it needs a generated cpp file.

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb commented:
@jgoppert posix build now gets to:
[ 44%] Building C object src/systemcmds/ver/CMakeFiles/ver.dir/ver.c.o
In file included from /local2/mnt/workspace/dronecode/Firmware/src/systemcmds/ver/ver.c:49:0:
/local2/mnt/workspace/dronecode/Firmware/src/modules/systemlib/git_version.h:44:31: fatal error: build_git_version.h: No such file or directory
 #include "build_git_version.h"

mcharleb commented:
Workaround:
$ cat src/include/build_git_version.h 
#define PX4_GIT_VERSION_STR    "foo"
#define PX4_GIT_VERSION_BINARY "bar"

jgoppert commented:
We need to convert this from the original makefile.
```bash
#
# Versioning
#

GIT_VER_FILE = $(PX4_VERSIONING_DIR).build_git_ver
GIT_HEADER_FILE = $(PX4_VERSIONING_DIR)build_git_version.h

$(GIT_VER_FILE) :
	$(Q) if [ ! -f $(GIT_VER_FILE) ]; then \
		$(MKDIR) -p $(PX4_VERSIONING_DIR); \
		$(ECHO) "" > $(GIT_VER_FILE); \
	fi

.PHONY: checkgitversion
checkgitversion: $(GIT_VER_FILE)
	$(Q) if [ "$(GIT_DESC)" != "$(shell cat $(GIT_VER_FILE))" ]; then \
		$(ECHO) "/* Auto Magically Generated file */" > $(GIT_HEADER_FILE); \
		$(ECHO) "/* Do not edit! */" >> $(GIT_HEADER_FILE); \
		$(ECHO) "#define PX4_GIT_VERSION_STR  \"$(GIT_DESC)\"" >> $(GIT_HEADER_FILE); \
		$(ECHO) "#define PX4_GIT_VERSION_BINARY 0x$(GIT_DESC_SHORT)" >> $(GIT_HEADER_FILE); \
		$(ECHO) $(GIT_DESC) > $(GIT_VER_FILE); \
	fi
#

```

mcharleb committed:
Added cmake support for qurt

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
Fixed CMakeLIsts.txt for topic_listener

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
cmake: add git version header file generation

Uses src/include/build_git_version.h.in as a template to generate
build_git_version.h in the build dir

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb commented:
The header file generation has been added. It still needs to be reworked to generate the GIT_DESC but the rest is automated.

mcharleb committed:
cmake: get the git version using execute_process

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb commented:
@jgoppert GIT_DESC is not set dynamically

mcharleb commented:
@LorenzMeier posix and qurt now fully build. Still need to verify flags are set correctly.

LorenzMeier commented:
@jgoppert Could we do a tiny, tiny NuttX config and see if that builds and runs? Or did I miss that we reached this already?

jgoppert commented:
@LorenzMeier I can do that. I had it compiling but didn't try uploading it yet to a board. I think the recent changes made broke the nuttx build, but worst case I can roll back to when I had it working. The nuttx still needs to get the builtins and romfs generated, but that shouldn't be too bad.

jgoppert commented:
I fixed the linking, now I just have to finish packaging the firmware.
